### ISSUE TITLE
Trigger loop encountered using DailyTimeIntervalTrigger across DST start boundary

### ISSUE THREAD

**Issue by dandufeu** [quartznet/quartznet#332]
I encountered an issue that just occurred on the last Daylight Savings Time (DST) change on March 12, 2016.  I'm using a DailyTimeIntervalTrigger that was set to fire before, during and after the DST change. The triggers are set to run in various time zones (according to location). It appears that about an **hour** after the DST change, the trigger keeps firing without pause (like it is in an infinite loop).

A simplified version of the code I'm using for the trigger is as follows:

``` c#
TimeZoneInfo tz = TimeZoneInfo.Local;
ITrigger trigger = TriggerBuilder.Create()
    .WithIdentity("MyJob")
    .WithDailyTimeIntervalSchedule(
        x => x.WithIntervalInMinutes(5)
            .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0)) 
            .EndingDailyAt(TimeOfDay.HourAndMinuteOfDay(23, 59))
            .OnDaysOfTheWeek(new DayOfWeek[] { DayOfWeek.Sunday, DayOfWeek.Monday, DayOfWeek.Tuesday, DayOfWeek.Wednesday, DayOfWeek.Thursday, DayOfWeek.Friday, DayOfWeek.Saturday })
            .InTimeZone(tz)
).Build();
```

I've Attached a sample project [QuartzTest.zip](https://github.com/quartznet/quartznet/files/173131/QuartzTest.zip)
If you set your clock to ~30 minutes before the DST start, and run it, you find that the program enters a tight loop of triggers an hour after DST changes. 

I put in some debug code in the test sample that prints out the result of ComputeFireTimes. I noticed that there was repetition after the DST start (I'm in Pacific Time, so DST start is at 10 AM UTC). I'm pretty sure that the repetition in the computation is the part of the problem. I haven't been able to investigate further yet. I'm testing with the 2.3.3 stable.

> 3/13/2016 9:35:00 AM +00:00
> 3/13/2016 9:40:00 AM +00:00
> 3/13/2016 9:45:00 AM +00:00
> 3/13/2016 9:50:00 AM +00:00
> 3/13/2016 9:55:00 AM +00:00
> 3/13/2016 10:00:00 AM +00:00
> 3/13/2016 10:05:00 AM +00:00
> 3/13/2016 10:10:00 AM +00:00
> 3/13/2016 10:15:00 AM +00:00
> 3/13/2016 10:20:00 AM +00:00
> 3/13/2016 10:25:00 AM +00:00
> 3/13/2016 10:30:00 AM +00:00
> 3/13/2016 10:35:00 AM +00:00
> 3/13/2016 10:40:00 AM +00:00
> 3/13/2016 10:45:00 AM +00:00
> 3/13/2016 10:50:00 AM +00:00
> 3/13/2016 10:55:00 AM +00:00
> 3/13/2016 10:00:00 AM +00:00
> 3/13/2016 10:05:00 AM +00:00
> 3/13/2016 10:10:00 AM +00:00
> 3/13/2016 10:15:00 AM +00:00
> 3/13/2016 10:20:00 AM +00:00
> 3/13/2016 10:25:00 AM +00:00
> 3/13/2016 10:30:00 AM +00:00
> 3/13/2016 10:35:00 AM +00:00
> 3/13/2016 10:40:00 AM +00:00
> 3/13/2016 10:45:00 AM +00:00
> 3/13/2016 10:50:00 AM +00:00
> 3/13/2016 10:55:00 AM +00:00
> 3/13/2016 10:00:00 AM +00:00
> 3/13/2016 10:05:00 AM +00:00
> 3/13/2016 10:10:00 AM +00:00
> 3/13/2016 10:15:00 AM +00:00
> 3/13/2016 10:20:00 AM +00:00
> 3/13/2016 10:25:00 AM +00:00
> 3/13/2016 10:30:00 AM +00:00
> 3/13/2016 10:35:00 AM +00:00
> 3/13/2016 10:40:00 AM +00:00
> 3/13/2016 10:45:00 AM +00:00
> 3/13/2016 10:50:00 AM +00:00
> 3/13/2016 10:55:00 AM +00:00
> 3/13/2016 10:00:00 AM +00:00
> 3/13/2016 10:05:00 AM +00:00
> 3/13/2016 10:10:00 AM +00:00
> 3/13/2016 10:15:00 AM +00:00
> 3/13/2016 10:20:00 AM +00:00
> 3/13/2016 10:25:00 AM +00:00
> 3/13/2016 10:30:00 AM +00:00
> 3/13/2016 10:35:00 AM +00:00
> 3/13/2016 10:40:00 AM +00:00

**Comment by rrajewski** [comment]
We had the same problem yesterday with any jobs that either had a StartDailyAt specified before 3AM or were scheduled OnEveryDay without a StartDailyAt.

`(s =>
     s.WithIntervalInMinutes(5)
     .OnEveryDay()
     .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0))
)`
or
`(s =>
     s.WithIntervalInMinutes(5)
     .OnEveryDay()
 )
`

**Comment by TyBarthel** [comment]
We had this occur again too.  It's creating some major headaches on every DST change.

**Comment by lahma** [comment]
Unfortunately these DST problems still seem to affect also 2.5 release.

**Comment by rrajewski** [comment]
It is still a problem for me with version 2.5.

If I change my computer's date back to 3/12/17, the job runs in an infinite loop. I also updated my trigger to match yours and got the same result:

            IScheduler scheduler = StdSchedulerFactory.GetDefaultScheduler();
            scheduler.Start();

            IJobDetail job = JobBuilder.Create<ProcessNewWorkflowsJob>().Build();

            var tz = TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time");
            ITrigger trigger = TriggerBuilder.Create()
                .StartAt(new DateTimeOffset(2016, 3, 13, 9, 0, 0, TimeSpan.Zero))
                .WithDailyTimeIntervalSchedule(
                x => x.WithIntervalInMinutes(5)
                    .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0))
                    .EndingDailyAt(TimeOfDay.HourAndMinuteOfDay(23, 59))
                    .OnDaysOfTheWeek(new DayOfWeek[] { DayOfWeek.Sunday, DayOfWeek.Monday, DayOfWeek.Tuesday, DayOfWeek.Wednesday, DayOfWeek.Thursday, DayOfWeek.Friday, DayOfWeek.Saturday })
                    .InTimeZone(tz)
            ).Build();

            scheduler.ScheduleJob(job, trigger);

**Comment by brandondahler** [comment]
Also had this (or a similar) issue.  Was able to reproduce on v2.4.1, v2.5.0, and [master@98e37323c9289c31afc92a2caa203664f6da8b74](https://github.com/quartznet/quartznet/tree/98e37323c9289c31afc92a2caa203664f6da8b74) (after fixing nuget package issue).

The general issue seems to be that when the when trying to resolve the next fire time after 2017-03-12 09:00:00 +00:00 when the trigger is in the Central Standard Timezone, it again resolve the next fire time to 2017-03-12 09:00:00 +00:00.  I'm sure the issue is also applicable to other DST timezones, but I did not explore those cases further.

Patch for a reproduction test below.

``` diff
diff --git a/src/Quartz.Tests.Unit/Impl/Triggers/DailyTimeIntervalTriggerImplTest.cs b/src/Quartz.Tests.Unit/Impl/Triggers/DailyTimeIntervalTriggerImplTest.cs
index 4cae2bc..9e7b32a 100644
--- a/src/Quartz.Tests.Unit/Impl/Triggers/DailyTimeIntervalTriggerImplTest.cs
+++ b/src/Quartz.Tests.Unit/Impl/Triggers/DailyTimeIntervalTriggerImplTest.cs
@@ -846,6 +846,22 @@ public void TestDayLightSaving()
         }
 
         [Test]
+        public void TestDayLightSaving2()
+        {
+            var timeZoneInfo = TimeZoneInfo.FindSystemTimeZoneById("Central Standard Time");
+
+            var trigger = DailyTimeIntervalScheduleBuilder.Create()
+                .OnEveryDay()
+                .InTimeZone(timeZoneInfo)
+                .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0))
+                .WithIntervalInHours(4)
+                .Build();
+
+            var first = trigger.GetFireTimeAfter(new DateTimeOffset(2017, 3, 12, 9, 0, 0, TimeSpan.Zero));
+            Assert.That(first, !Is.EqualTo(new DateTimeOffset(2017, 3, 12, 9, 0, 0, TimeSpan.Zero)));
+        }
+
+        [Test]
         public void TestPassingMidnight()
         {
             IOperableTrigger trigger = (IOperableTrigger) DailyTimeIntervalScheduleBuilder.Create()

```

To apply the diff, save the contents above as QuartzBug.diff in the root directory and run the command `git apply QuartzBug.diff`


If applying to master and receive a nuget package error, you can apply this diff to fix that temporarily:
``` diff
diff --git a/src/Quartz.Tests.Integration/packages.config b/src/Quartz.Tests.Integration/packages.config
index 54ad00c..2559241 100644
--- a/src/Quartz.Tests.Integration/packages.config
+++ b/src/Quartz.Tests.Integration/packages.config
@@ -4,6 +4,6 @@
   <package id="Common.Logging.Core" version="3.0.0" targetFramework="net40" />
   <package id="FirebirdSql.Data.FirebirdClient" version="4.5.0.0" targetFramework="net40" />
   <package id="NUnit" version="2.6.3" targetFramework="net40" />
-  <package id="odp.net.managed" version="121.1.1" targetFramework="net40" />
+  <package id="Oracle.ManagedDataAccess" version="12.1.24160719" targetFramework="net40" />
   <package id="RhinoMocks" version="3.6.1" />
 </packages>
\ No newline at end of file
diff --git a/src/Quartz.Tests.Unit/packages.config b/src/Quartz.Tests.Unit/packages.config
index dbcf6c8..eae9dc9 100644
--- a/src/Quartz.Tests.Unit/packages.config
+++ b/src/Quartz.Tests.Unit/packages.config
@@ -3,6 +3,6 @@
   <package id="Common.Logging" version="3.0.0" targetFramework="net40" />
   <package id="Common.Logging.Core" version="3.0.0" targetFramework="net40" />
   <package id="NUnit" version="2.6.3" targetFramework="net40" />
-  <package id="odp.net.managed" version="121.1.1" targetFramework="net40" />
+  <package id="Oracle.ManagedDataAccess" version="12.1.24160719" targetFramework="net40" />
   <package id="RhinoMocks" version="3.6.1" />
 </packages>
\ No newline at end of file
```

**Comment by Danielku15** [comment]
We also had serious problems with this just over this weekend where the DST was changed again. All installations of our product suddenly were running in an endless loop causing 100% CPU load and flooding the disk with log entries. 

We are using this kind of schedule:

```csharp
var jobTrigger = TriggerBuilder.Create()
    .WithDailyTimeIntervalSchedule(s => 
        s.OnEveryDay()
        .StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0))
        .WithIntervalInMinutes(10)
    .Build()
;
```

We were using version `2.4.1` but according to the history of this issue `2.5` is also affected. Is there already a known workaround or fix in sight?

**Comment by dandufeu** [comment]
@Danielku15 - You could try my fix that I pushed in my fork mentioned above. I had no reported issues during the latest DST changes.
https://github.com/dandufeu/quartznet/commit/925c3c86ad494b94c3a26b9d2cca9ad206b20aa6

**Comment by lahma** [comment]
@dandufeu 's fix looks valid and I have applied it, will be in 2.6. Thanks for helping out!

**Comment by Wojtekxor2** [comment]
I still use 2.6.0 version. The day after time change (25.03.2018) my job was run just like if it was in a loop. It starts immediately after configured.
My post is here: https://stackoverflow.com/questions/49477074/quartz-net-fires-every-second-in-the-spring-time-after-time-change

**Comment by lahma** [comment]
Reopening: this regressed and is being fixed again in #3190 (3.x) and #3191 (main).

The 2017 fix was correct, but the fall-back fix in #2917 / #2921 later added a date correction that applied to every interval size in either direction. On a spring-forward day — a 23 hour local day — it moved the computed fire time *backwards* into the previous local day, so `GetFireTimeAfter` returned a time at or before the one it was given and the scheduler span exactly as originally described here. Every `DailyTimeIntervalTrigger` with a repeat interval of an hour or less was affected, in every DST time zone, including the plain `WithIntervalInMinutes(n)` + `OnEveryDay()` shape reported in this thread. That correction is now scoped to triggers that fire at most once per local day, and can no longer move a fire time backwards.

A second cause, which predates all of it: resetting the time of day to `startTimeOfDay` when rolling over to the next day inherited whatever UTC offset the fire time already carried. In time zones that move the clock at midnight, `startTimeOfDay` 00:00 does not exist on the transition day and resolved to an instant before the transition — the same instant passed in. Also fixed.

The test added here in 2017, `TestDayLightSaving2`, never set `StartTimeUtc`, so it asserted against a clamped "now" and could not fail. That is why the regression shipped green. It now sets an explicit start time and asserts the exact expected fire time, alongside regression tests covering 3024 time zone / interval / start-time combinations.

Affected releases: 3.17.0 through 3.19.0.
